src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -928,19 +928,91 @@ DeferErrorIfCannotPushdownSubquery(Query *subqueryTree, bool outerMostQueryHasLi
 					  "functions";
 	}
 
-	if (subqueryTree->limitOffset)
+	/*
+	 * Correlated subqueries are effectively functions that are repeatedly called
+	 * for the values of the vars that point to the outer query. We can liberally
+	 * push down SQL features within such a function, as long as co-located join
+	 * checks are applied.
+	 */
+	if (!contain_vars_of_level((Node *) subqueryTree, 1))
```

Comment:
Looking into this check and `ContainsReferencesToOuterQueryWalker()`, I can see some differences. It looks like using `ContainsReferencesToOuterQuery()`  could help us support multiple levels of nested subqueries such as below, and could make it easier to read the code.


My understanding is that the logic explained on the PR description should hold for any var-level, right? The explain plans also verifies this.

```SQL
SELECT *
FROM users_table u
WHERE EXISTS
    (SELECT 1
     FROM events_table e
     WHERE u.user_id = e.user_id
       AND EXISTS
         (SELECT 1
          FROM users_table u2
          WHERE u2.user_id = u.user_id
          LIMIT 1));
```